src/jdk.incubator.code/share/classes/jdk/incubator/code/type/CoreTypeFactory.java

```diff
@@ -94,163 +97,8 @@ public TypeElement constructType(TypeElement.ExternalizedTypeElement tree) {
-
-                    List<RecordTypeRef.ComponentRef> cs = new ArrayList<>();
-                    for (int i = 1; i < tree.arguments().size(); i += 2) {
-                        ExternalizedTypeElement cname = tree.arguments().get(i + 1);
-                        if (!cname.arguments().isEmpty()) {
-                            throw badType(tree, "record reference");
-                        }
-                        cs.add(new RecordTypeRef.ComponentRef(
-                                constructType(tree.arguments().get(i)),
-                                cname.identifier()));
-                    }
-                    return RecordTypeRef.recordType(constructType(tree.arguments().get(0)), cs);
-                } else {
-                    throw badType(tree, "unknown reference");
-                }
-            } else if (identifier.equals(".")) {
-                // qualified type
-                if (tree.arguments().size() != 2) {
-                    throw badType(tree, "qualified type");
-                }
-                ClassType enclType = (ClassType)constructTypeArgument(tree, 0, CLASS);
-                ClassType innerType = (ClassType)constructTypeArgument(tree, 1, CLASS);
-                // the inner class name is obtained by subtracting the name of the enclosing type
-                // from the name of the inner type (and also dropping an extra '$')
-                String innerName = innerType.toNominalDescriptor().displayName()
-                        .substring(enclType.toNominalDescriptor().displayName().length() + 1);
-                JavaType qual = JavaType.qualified(enclType, innerName);
-                return (innerType.hasTypeArguments()) ?
-                    JavaType.parameterized(qual, innerType.typeArguments()) : qual;
-            } else {
-                // primitive or reference
-                JavaType t = switch (identifier) {
-                    case "boolean" -> JavaType.BOOLEAN;
-                    case "byte" -> JavaType.BYTE;
-                    case "char" -> JavaType.CHAR;
-                    case "short" -> JavaType.SHORT;
-                    case "int" -> JavaType.INT;
-                    case "long" -> JavaType.LONG;
-                    case "float" -> JavaType.FLOAT;
-                    case "double" -> JavaType.DOUBLE;
-                    case "void" -> JavaType.VOID;
-                    default -> JavaType.type(ClassDesc.of(identifier));
-                };
-                if (!tree.arguments().isEmpty()) {
-                    if (t instanceof PrimitiveType) {
-                        throw new IllegalArgumentException("primitive type: " + tree);
-                    }
-                    return JavaType.parameterized(t,
-                            tree.arguments().stream().map(ete -> (JavaType) constructType(ete)).toList());
-                } else {
-                    return t;
-                }
-            }
+            return JavaTypeUtils.toJavaType(tree);
```

Comment:
Note how parsing an inflated type form is now done entirely inside `JavaTypeUtils`.

I'm not entirely sure whether the Java type factory should support both Java type and reference inflated forms. For now I went with just types -- but if we want to push this further (e.g. to cover also op attributes) we would need the factory to support both types and refs.